arch/x86/kvm/vmx/pkvm/hyp/pci.c

```diff
@@ -52,6 +52,18 @@ static int pci_mmcfg_write(u64 address, int size, unsigned long value)
 	return 0;
 }
 
+static bool host_vpci_cfg_space_audit_access(struct pkvm_ptdev *ptdev, u64 offset)
+{
+	if (!ptdev || ptdev->shadow_vm_handle == 0)
+		return true;
+
+	/* Deny for all BAR access */
+	if (offset >= 0x10 && offset < 0x28)
+		return false;
```

Comment:
I suppose that in the final solution we will need audit only for BAR write, not for BAR read? Maybe let's not deny BAR read access?

BTW, with this PR, I ran a Linux VM via crosvm as a protected VM with iwlwifi passed-thru, and iwlwifi was failing in the VM:
```
[    1.035269] iwlwifi 0000:00:14.0: HW_REV=0xFFFFFFFF, PCI issues?
[    1.035671] iwlwifi: probe of 0000:00:14.0 failed with error -5
```

Even after I commented out audit for BAR read, it still wasn't working. Only after I commented out audit for BAR write as well, iwlwifi started working again.

So at least in some setups this PR makes a protected VM's passthrough device non-workable (i.e. the VMM needs access to physical BARs even after attaching the device to the VM).

Reply:
For vfio-pci pass-thru, the BAR should be totally emulated, it means guest BAR program should not finally touch real ones. We need ensure what's the behavior host will do for BAR programing after assign a device.

Reply:
Yep, I know, I meant BAR access by the host (which, though, apparently affects the guest's emulated PCI cfg space functionality too). I was observing a few dozen `pkvm: handle host port I/O access failed` errors at the moment of the VM startup. I can debug why exactly crosvm does that.

Reply:
During VM startup, BAR reading from host is possible, but what's the reason of BAR writing?

Reply:
I updated the patchset to allow BAR read.

For the BAR writing behavior dmytro you notice, I believe this is done when crosvm create the vfio device:
crosvm call VFIO_GROUP_GET_DEVICE_FD ->
vfio_group_fops_unl_ioctl->
vfio_group_get_device_fd->
vfio_device_open->
vfio_pci_open_device->
vfio_pci_core_enable->
pci_try_reset_function->
pci_dev_restore

So the crosvm first call KVM_CREATE_DEVICE, where pKVM add the device to shadow vm.
Then crosvm calls VFIO_GROUP_GET_DEVICE_FD, which would reset the pci function and trigger BAR write.

I make a test to reverse the two in crosvm, ie. first get the vfio fd then create the kvm device, it works. No BAR write is seen after the kvm device is created.
Maybe dmytro you could have a try with this?

Reply:
@tzhang449 , do you mean BAR reading instead of BAR writing here for VFIO_GROUP_GET_DEVICE_FD?

> So the crosvm first call KVM_CREATE_DEVICE, where pKVM add the device to shadow vm. Then crosvm calls VFIO_GROUP_GET_DEVICE_FD, which would reset the pci function and trigger BAR write.
> 
> I make a test to reverse the two in crosvm, ie. first get the vfio fd then create the kvm device, it works. No BAR write is seen after the kvm device is created. Maybe dmytro you could have a try with this?

Reply:
Agree, I will try cache the BAR value when pkvm attach ptdev and audit the BAR write based this.

Reply:
Dmytro thanks for your feedback. I updated the patch to cache the BAR value when pkvm attach ptdev, and allow BAR write only if it is writing the cached value into the BAR. Looks like things are ok for now.

Reply:
Thanks. I confirm that now it works ok on my side too.

Reply:
There is still `/* Protects updates of pgt pointer */` comment in `ptdev.h`, it probably should be removed since now ptdev->lock is used not only for ptdev->pgt updates.

Reply:
Agree, will remove this.

Reply:
Removed the comment  in latest branch.